// @clr/angular/esm2015/utils/drag-and-drop/providers/drag-handle-registrar.service.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable, Renderer2 } from '@angular/core';
import { DragEventListenerService } from './drag-event-listener.service';
// This provider registers the drag handle element.
// When it registers a element as a drag handle, it attaches that element to the listeners from ClrDragEventListener.
// Also, it adds the "drag-handle" css class to the registered element through Renderer.
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './drag-event-listener.service';
export class DragHandleRegistrarService {
    constructor(dragEventListener, renderer) {
        this.dragEventListener = dragEventListener;
        this.renderer = renderer;
    }
    get defaultHandleEl() {
        return this._defaultHandleEl;
    }
    set defaultHandleEl(el) {
        this._defaultHandleEl = el; // defaultHandleEl will be usually the clrDraggable element.
        // If the customHandleEl has been registered,
        // don't make the defaultHandleEl the drag handle yet until the customHandleEl is unregistered.
        if (!this._customHandleEl) {
            this.makeElementHandle(this._defaultHandleEl);
        }
    }
    makeElementHandle(el) {
        if (this._defaultHandleEl && this._defaultHandleEl !== el) {
            // Before making an element the custom handle element,
            // we should remove the existing drag-handle class from the draggable element.
            this.renderer.removeClass(this._defaultHandleEl, 'drag-handle');
        }
        this.dragEventListener.attachDragListeners(el);
        this.renderer.addClass(el, 'drag-handle');
    }
    get customHandleEl() {
        return this._customHandleEl;
    }
    registerCustomHandle(el) {
        this.dragEventListener.detachDragListeners(); // removes the existing listeners
        this._customHandleEl = el;
        this.makeElementHandle(this._customHandleEl);
    }
    unregisterCustomHandle() {
        this.dragEventListener.detachDragListeners(); // removes the existing listeners
        this.renderer.removeClass(this._customHandleEl, 'drag-handle');
        delete this._customHandleEl;
        // if default handle is set, make that handle
        if (this._defaultHandleEl) {
            this.makeElementHandle(this._defaultHandleEl);
        }
    }
}
DragHandleRegistrarService.ɵfac = function DragHandleRegistrarService_Factory(t) { return new (t || DragHandleRegistrarService)(ɵngcc0.ɵɵinject(ɵngcc1.DragEventListenerService), ɵngcc0.ɵɵinject(ɵngcc0.Renderer2)); };
DragHandleRegistrarService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: DragHandleRegistrarService, factory: DragHandleRegistrarService.ɵfac });
DragHandleRegistrarService.ctorParameters = () => [
    { type: DragEventListenerService },
    { type: Renderer2 }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DragHandleRegistrarService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.DragEventListenerService }, { type: ɵngcc0.Renderer2 }]; }, null); })();
