// @clr/angular/esm2015/utils/drag-and-drop/providers/draggable-snapshot.service.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable } from '@angular/core';
import { DomAdapter } from '../../dom-adapter/dom-adapter';
// This service is used to capture the state of clrDraggable element
// at a certain event and passes it to clrDraggableGhost component.
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../dom-adapter/dom-adapter';
export class DraggableSnapshotService {
    constructor(domAdapter) {
        this.domAdapter = domAdapter;
    }
    capture(el, event) {
        this.draggableElClientRect = this.domAdapter.clientRect(el);
        this.snapshotDragEvent = event;
    }
    discard() {
        delete this.draggableElClientRect;
        delete this.snapshotDragEvent;
    }
    get hasDraggableState() {
        return !!this.snapshotDragEvent && !!this.draggableElClientRect;
    }
    get clientRect() {
        return this.draggableElClientRect;
    }
    get dragEvent() {
        return this.snapshotDragEvent;
    }
}
DraggableSnapshotService.ɵfac = function DraggableSnapshotService_Factory(t) { return new (t || DraggableSnapshotService)(ɵngcc0.ɵɵinject(ɵngcc1.DomAdapter)); };
DraggableSnapshotService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: DraggableSnapshotService, factory: DraggableSnapshotService.ɵfac });
DraggableSnapshotService.ctorParameters = () => [
    { type: DomAdapter }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DraggableSnapshotService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.DomAdapter }]; }, null); })();
