// @clr/angular/esm2015/layout/vertical-nav/providers/vertical-nav-group-registration.service.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class VerticalNavGroupRegistrationService {
    constructor() {
        this.navGroupCount = 0;
    }
    registerNavGroup() {
        this.navGroupCount++;
    }
    unregisterNavGroup() {
        this.navGroupCount--;
    }
}
VerticalNavGroupRegistrationService.ɵfac = function VerticalNavGroupRegistrationService_Factory(t) { return new (t || VerticalNavGroupRegistrationService)(); };
VerticalNavGroupRegistrationService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: VerticalNavGroupRegistrationService, factory: VerticalNavGroupRegistrationService.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(VerticalNavGroupRegistrationService, [{
        type: Injectable
    }], function () { return []; }, null); })();
