// @clr/angular/esm2015/utils/animations/expandable-animation/expandable-animation.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EXPANDABLE_ANIMATION_DIRECTIVES } from './index';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './expandable-animation';
export class ClrExpandableAnimationModule {
}
ClrExpandableAnimationModule.ɵfac = function ClrExpandableAnimationModule_Factory(t) { return new (t || ClrExpandableAnimationModule)(); };
ClrExpandableAnimationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrExpandableAnimationModule });
ClrExpandableAnimationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrExpandableAnimationModule, { declarations: function () { return [ɵngcc1.ClrExpandableAnimation]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ɵngcc1.ClrExpandableAnimation]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrExpandableAnimationModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [EXPANDABLE_ANIMATION_DIRECTIVES],
                exports: [EXPANDABLE_ANIMATION_DIRECTIVES]
            }]
    }], null, null); })();
