// @clr/angular/esm2015/forms/common/if-control-state/abstract-if-state.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, Optional } from '@angular/core';
import { NgControlService } from '../providers/ng-control.service';
import { IfControlStateService } from './if-control-state.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './if-control-state.service';
import * as ɵngcc2 from '../providers/ng-control.service';
export class AbstractIfState {
    constructor(ifControlStateService, ngControlService) {
        this.ifControlStateService = ifControlStateService;
        this.ngControlService = ngControlService;
        this.subscriptions = [];
        this.displayedContent = false;
        if (ngControlService) {
            this.subscriptions.push(this.ngControlService.controlChanges.subscribe(control => {
                this.control = control;
            }));
        }
        if (ifControlStateService) {
            this.subscriptions.push(this.ifControlStateService.statusChanges.subscribe((state) => {
                this.handleState(state);
            }));
        }
    }
    ngOnDestroy() {
        this.subscriptions.forEach(sub => sub.unsubscribe());
    }
    handleState(state) {
        /* overwrite in implementation to handle status change */
    }
}
AbstractIfState.ɵfac = function AbstractIfState_Factory(t) { return new (t || AbstractIfState)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.IfControlStateService, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.NgControlService, 8)); };
AbstractIfState.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AbstractIfState });
AbstractIfState.ctorParameters = () => [
    { type: IfControlStateService, decorators: [{ type: Optional }] },
    { type: NgControlService, decorators: [{ type: Optional }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AbstractIfState, [{
        type: Directive
    }], function () { return [{ type: ɵngcc1.IfControlStateService, decorators: [{
                type: Optional
            }] }, { type: ɵngcc2.NgControlService, decorators: [{
                type: Optional
            }] }]; }, null); })();
