// @clr/angular/esm2015/forms/common/if-control-state/if-success.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, Optional, TemplateRef, ViewContainerRef } from '@angular/core';
import { NgControlService } from '../providers/ng-control.service';
import { IfControlStateService, CONTROL_STATE } from './if-control-state.service';
import { AbstractIfState } from './abstract-if-state';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './if-control-state.service';
import * as ɵngcc2 from '../providers/ng-control.service';
export class ClrIfSuccess extends AbstractIfState {
    constructor(ifControlStateService, ngControlService, template, container) {
        super(ifControlStateService, ngControlService);
        this.template = template;
        this.container = container;
        if (!ifControlStateService) {
            throw new Error('ClrIfSuccess can only be used within a form control container element like clr-input-container');
        }
    }
    /**
     * @param state CONTROL_STATE
     */
    handleState(state) {
        const isValid = CONTROL_STATE.VALID === state;
        if (isValid && !this.displayedContent) {
            this.container.createEmbeddedView(this.template);
        }
        else if (!isValid && this.container) {
            this.container.clear();
        }
        this.displayedContent = isValid;
    }
}
ClrIfSuccess.ɵfac = function ClrIfSuccess_Factory(t) { return new (t || ClrIfSuccess)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.IfControlStateService, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.NgControlService, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
ClrIfSuccess.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrIfSuccess, selectors: [["", "clrIfSuccess", ""]], features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
ClrIfSuccess.ctorParameters = () => [
    { type: IfControlStateService, decorators: [{ type: Optional }] },
    { type: NgControlService, decorators: [{ type: Optional }] },
    { type: TemplateRef },
    { type: ViewContainerRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrIfSuccess, [{
        type: Directive,
        args: [{ selector: '[clrIfSuccess]' }]
    }], function () { return [{ type: ɵngcc1.IfControlStateService, decorators: [{
                type: Optional
            }] }, { type: ɵngcc2.NgControlService, decorators: [{
                type: Optional
            }] }, { type: ɵngcc0.TemplateRef }, { type: ɵngcc0.ViewContainerRef }]; }, null); })();
