// @clr/angular/esm2015/forms/common/providers/ng-control.service.js
/*
 * Copyright (c) 2016-2021 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
export class NgControlService {
    constructor() {
        // Observable to subscribe to the control, since its not available immediately for projected content
        this._controlChanges = new Subject();
        this._helpers = new Subject();
    }
    get controlChanges() {
        return this._controlChanges.asObservable();
    }
    setControl(control) {
        this._controlChanges.next(control);
    }
    get helpersChange() {
        return this._helpers.asObservable();
    }
    setHelpers(state) {
        this._helpers.next(state);
    }
}
NgControlService.ɵfac = function NgControlService_Factory(t) { return new (t || NgControlService)(); };
NgControlService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NgControlService, factory: NgControlService.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgControlService, [{
        type: Injectable
    }], function () { return []; }, null); })();
