// @clr/angular/esm2015/forms/textarea/textarea.module.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { checkCircleIcon, ClarityIcons, exclamationCircleIcon } from '@cds/core/icon';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrTextarea } from './textarea';
import { ClrTextareaContainer } from './textarea-container';
import * as ɵngcc0 from '@angular/core';
export class ClrTextareaModule {
    constructor() {
        ClarityIcons.addIcons(exclamationCircleIcon, checkCircleIcon);
    }
}
ClrTextareaModule.ɵfac = function ClrTextareaModule_Factory(t) { return new (t || ClrTextareaModule)(); };
ClrTextareaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrTextareaModule });
ClrTextareaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule], ClrCommonFormsModule] });
ClrTextareaModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrTextareaModule, { declarations: function () { return [ClrTextarea, ClrTextareaContainer]; }, imports: function () { return [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule]; }, exports: function () { return [ClrCommonFormsModule, ClrTextarea, ClrTextareaContainer]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrTextareaModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule],
                declarations: [ClrTextarea, ClrTextareaContainer],
                exports: [ClrCommonFormsModule, ClrTextarea, ClrTextareaContainer],
                entryComponents: [ClrTextareaContainer]
            }]
    }], function () { return []; }, null); })();
