// @clr/angular/esm2015/forms/range/range.module.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { checkCircleIcon, ClarityIcons, exclamationCircleIcon } from '@cds/core/icon';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrHostWrappingModule } from '../../utils/host-wrapping/host-wrapping.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrRange } from './range';
import { ClrRangeContainer } from './range-container';
import * as ɵngcc0 from '@angular/core';
export class ClrRangeModule {
    constructor() {
        ClarityIcons.addIcons(exclamationCircleIcon, checkCircleIcon);
    }
}
ClrRangeModule.ɵfac = function ClrRangeModule_Factory(t) { return new (t || ClrRangeModule)(); };
ClrRangeModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrRangeModule });
ClrRangeModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, ClrCommonFormsModule, ClrHostWrappingModule, ClrIconModule], ClrCommonFormsModule] });
ClrRangeModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrRangeModule, { declarations: function () { return [ClrRange, ClrRangeContainer]; }, imports: function () { return [CommonModule, ClrCommonFormsModule, ClrHostWrappingModule, ClrIconModule]; }, exports: function () { return [ClrCommonFormsModule, ClrRange, ClrRangeContainer]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrRangeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrCommonFormsModule, ClrHostWrappingModule, ClrIconModule],
                declarations: [ClrRange, ClrRangeContainer],
                exports: [ClrCommonFormsModule, ClrRange, ClrRangeContainer],
                entryComponents: [ClrRangeContainer]
            }]
    }], function () { return []; }, null); })();
