// @clr/angular/esm2015/utils/focus-trap/focus-trap.module.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FocusTrapDirective } from './focus-trap.directive';
import * as ɵngcc0 from '@angular/core';
export const FOCUS_TRAP_DIRECTIVES = [FocusTrapDirective];
export class ClrFocusTrapModule {
}
ClrFocusTrapModule.ɵfac = function ClrFocusTrapModule_Factory(t) { return new (t || ClrFocusTrapModule)(); };
ClrFocusTrapModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrFocusTrapModule });
ClrFocusTrapModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrFocusTrapModule, { declarations: function () { return [FocusTrapDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [FocusTrapDirective]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrFocusTrapModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [FOCUS_TRAP_DIRECTIVES],
                exports: [FOCUS_TRAP_DIRECTIVES]
            }]
    }], null, null); })();
