// angular-canvas/esm2015/lib/canvas-dom.module.js
import { APP_ID, NgModule, NgZone, NO_ERRORS_SCHEMA, RendererFactory2, } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CanvasDomRendererFactory } from './canvas-dom-renderer';
import { EventManager, ɵDomSharedStylesHost as DomSharedStylesHost, } from '@angular/platform-browser';
import * as ɵngcc0 from '@angular/core';
export class CanvasDomModule {
    // tslint:disable-next-line:typedef
    static forRoot(elements) {
        return {
            ngModule: CanvasDomModule,
        };
    }
}
CanvasDomModule.ɵfac = function CanvasDomModule_Factory(t) { return new (t || CanvasDomModule)(); };
CanvasDomModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: CanvasDomModule });
CanvasDomModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: RendererFactory2,
            useClass: CanvasDomRendererFactory,
            deps: [EventManager, DomSharedStylesHost, APP_ID, NgZone]
        },
    ], imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CanvasDomModule, { imports: function () { return [CommonModule]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CanvasDomModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                providers: [
                    {
                        provide: RendererFactory2,
                        useClass: CanvasDomRendererFactory,
                        deps: [EventManager, DomSharedStylesHost, APP_ID, NgZone]
                    },
                ],
                schemas: [NO_ERRORS_SCHEMA]
            }]
    }], null, null); })();
