// @clr/angular/esm2015/forms/radio/radio.module.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { checkCircleIcon, ClarityIcons, exclamationCircleIcon } from '@cds/core/icon';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrHostWrappingModule } from '../../utils/host-wrapping/host-wrapping.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrRadio } from './radio';
import { ClrRadioContainer } from './radio-container';
import { ClrRadioWrapper } from './radio-wrapper';
import * as ɵngcc0 from '@angular/core';
export class ClrRadioModule {
    constructor() {
        ClarityIcons.addIcons(exclamationCircleIcon, checkCircleIcon);
    }
}
ClrRadioModule.ɵfac = function ClrRadioModule_Factory(t) { return new (t || ClrRadioModule)(); };
ClrRadioModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrRadioModule });
ClrRadioModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, ClrCommonFormsModule, ClrHostWrappingModule, ClrIconModule], ClrCommonFormsModule] });
ClrRadioModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrRadioModule, { declarations: function () { return [ClrRadio, ClrRadioContainer, ClrRadioWrapper]; }, imports: function () { return [CommonModule, ClrCommonFormsModule, ClrHostWrappingModule, ClrIconModule]; }, exports: function () { return [ClrCommonFormsModule, ClrRadio, ClrRadioContainer, ClrRadioWrapper]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrRadioModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrCommonFormsModule, ClrHostWrappingModule, ClrIconModule],
                declarations: [ClrRadio, ClrRadioContainer, ClrRadioWrapper],
                exports: [ClrCommonFormsModule, ClrRadio, ClrRadioContainer, ClrRadioWrapper],
                entryComponents: [ClrRadioWrapper]
            }]
    }], function () { return []; }, null); })();
