// @clr/angular/esm2015/popover/dropdown/dropdown-trigger.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, ElementRef, HostListener } from '@angular/core';
import { ClrPopoverToggleService } from '../../utils/popover/providers/popover-toggle.service';
import { ClrDropdown } from './dropdown';
import { DropdownFocusHandler } from './providers/dropdown-focus-handler.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './dropdown';
import * as ɵngcc2 from '../../utils/popover/providers/popover-toggle.service';
import * as ɵngcc3 from './providers/dropdown-focus-handler.service';
export class ClrDropdownTrigger {
    constructor(dropdown, toggleService, el, focusHandler) {
        this.toggleService = toggleService;
        this.isRootLevelToggle = true;
        // if the containing dropdown has a parent, then this is not the root level one
        if (dropdown.parent) {
            this.isRootLevelToggle = false;
        }
        focusHandler.trigger = el.nativeElement;
    }
    get active() {
        return this.toggleService.open;
    }
    onDropdownTriggerClick(event) {
        this.toggleService.toggleWithEvent(event);
    }
}
ClrDropdownTrigger.ɵfac = function ClrDropdownTrigger_Factory(t) { return new (t || ClrDropdownTrigger)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ClrDropdown), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ClrPopoverToggleService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.DropdownFocusHandler)); };
ClrDropdownTrigger.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrDropdownTrigger, selectors: [["", "clrDropdownTrigger", ""], ["", "clrDropdownToggle", ""]], hostVars: 10, hostBindings: function ClrDropdownTrigger_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ClrDropdownTrigger_click_HostBindingHandler($event) { return ctx.onDropdownTriggerClick($event); });
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-haspopup", "menu")("aria-expanded", ctx.active);
        ɵngcc0.ɵɵclassProp("dropdown-toggle", ctx.isRootLevelToggle)("dropdown-item", !ctx.isRootLevelToggle)("expandable", !ctx.isRootLevelToggle)("active", ctx.active);
    } } });
ClrDropdownTrigger.ctorParameters = () => [
    { type: ClrDropdown },
    { type: ClrPopoverToggleService },
    { type: ElementRef },
    { type: DropdownFocusHandler }
];
ClrDropdownTrigger.propDecorators = {
    onDropdownTriggerClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrDropdownTrigger, [{
        type: Directive,
        args: [{
                // We support both selectors for legacy reasons
                selector: '[clrDropdownTrigger],[clrDropdownToggle]',
                host: {
                    '[class.dropdown-toggle]': 'isRootLevelToggle',
                    '[class.dropdown-item]': '!isRootLevelToggle',
                    '[class.expandable]': '!isRootLevelToggle',
                    '[class.active]': 'active',
                    '[attr.aria-haspopup]': '"menu"',
                    '[attr.aria-expanded]': 'active'
                }
            }]
    }], function () { return [{ type: ɵngcc1.ClrDropdown }, { type: ɵngcc2.ClrPopoverToggleService }, { type: ɵngcc0.ElementRef }, { type: ɵngcc3.DropdownFocusHandler }]; }, { onDropdownTriggerClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
