// @clr/angular/esm2015/forms/common/success.js
/**
 * Copyright (c) 2016-2021 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, Optional } from '@angular/core';
import { ClrAbstractControl, CONTROL_SUFFIX } from './abstract-control';
import { ContainerIdService } from './providers/container-id.service';
import { ControlIdService } from './providers/control-id.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './providers/control-id.service';
import * as ɵngcc2 from './providers/container-id.service';

const _c0 = ["*"];
export class ClrControlSuccess extends ClrAbstractControl {
    constructor(controlIdService, containerIdService) {
        super(controlIdService, containerIdService);
        this.controlIdService = controlIdService;
        this.containerIdService = containerIdService;
        this.controlIdSuffix = CONTROL_SUFFIX.SUCCESS;
    }
}
ClrControlSuccess.ɵfac = function ClrControlSuccess_Factory(t) { return new (t || ClrControlSuccess)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ControlIdService, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ContainerIdService, 8)); };
ClrControlSuccess.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrControlSuccess, selectors: [["clr-control-success"]], hostVars: 3, hostBindings: function ClrControlSuccess_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("id", ctx.id);
        ɵngcc0.ɵɵclassProp("clr-subtext", true);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrControlSuccess_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ClrControlSuccess.ctorParameters = () => [
    { type: ControlIdService, decorators: [{ type: Optional }] },
    { type: ContainerIdService, decorators: [{ type: Optional }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrControlSuccess, [{
        type: Component,
        args: [{
                selector: 'clr-control-success',
                template: ` <ng-content></ng-content> `,
                host: {
                    '[class.clr-subtext]': 'true',
                    '[attr.id]': 'id'
                }
            }]
    }], function () { return [{ type: ɵngcc1.ControlIdService, decorators: [{
                type: Optional
            }] }, { type: ɵngcc2.ContainerIdService, decorators: [{
                type: Optional
            }] }]; }, null); })();
