// @clr/angular/esm2015/utils/focus/key-focus/key-focus.module.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrKeyFocus } from './key-focus';
import { ClrKeyFocusItem } from './key-focus-item';
import { ClrRovingTabindex } from './roving-tabindex';
import * as ɵngcc0 from '@angular/core';
const KEY_FOCUS_DIRECTIVES = [ClrKeyFocus, ClrRovingTabindex, ClrKeyFocusItem];
export class ClrKeyFocusModule {
}
ClrKeyFocusModule.ɵfac = function ClrKeyFocusModule_Factory(t) { return new (t || ClrKeyFocusModule)(); };
ClrKeyFocusModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrKeyFocusModule });
ClrKeyFocusModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrKeyFocusModule, { declarations: function () { return [ClrKeyFocus, ClrRovingTabindex, ClrKeyFocusItem]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrKeyFocus, ClrRovingTabindex, ClrKeyFocusItem]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrKeyFocusModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [KEY_FOCUS_DIRECTIVES],
                exports: [KEY_FOCUS_DIRECTIVES]
            }]
    }], null, null); })();
