// @clr/angular/esm2015/forms/combobox/combobox.module.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrConditionalModule } from '../../utils/conditional/conditional.module';
import { ClrOption } from './option';
import { ClrOptions } from './options';
import { ClrCombobox } from './combobox';
import { ClrComboboxContainer } from './combobox-container';
import { ClrOptionSelected } from './option-selected.directive';
import { ClrPopoverModuleNext } from '../../utils/popover/popover.module';
import { ClrOptionItems } from './option-items.directive';
import { ClrSpinnerModule } from '../../progress/spinner/spinner.module';
import { ClrKeyFocusModule } from '../../utils/focus/key-focus/key-focus.module';
import { angleIcon, checkCircleIcon, ClarityIcons, exclamationCircleIcon, windowCloseIcon } from '@cds/core/icon';
import * as ɵngcc0 from '@angular/core';
export class ClrComboboxModule {
    constructor() {
        ClarityIcons.addIcons(exclamationCircleIcon, checkCircleIcon, angleIcon, windowCloseIcon);
    }
}
ClrComboboxModule.ɵfac = function ClrComboboxModule_Factory(t) { return new (t || ClrComboboxModule)(); };
ClrComboboxModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrComboboxModule });
ClrComboboxModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            ClrIconModule,
            ClrKeyFocusModule,
            ClrCommonFormsModule,
            ClrConditionalModule,
            ClrPopoverModuleNext,
            ClrSpinnerModule,
        ], ClrCommonFormsModule,
        ClrConditionalModule] });
ClrComboboxModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrComboboxModule, { declarations: function () { return [ClrCombobox, ClrComboboxContainer, ClrOptions, ClrOption, ClrOptionSelected, ClrOptionItems]; }, imports: function () { return [CommonModule,
        FormsModule,
        ClrIconModule,
        ClrKeyFocusModule,
        ClrCommonFormsModule,
        ClrConditionalModule,
        ClrPopoverModuleNext,
        ClrSpinnerModule]; }, exports: function () { return [ClrCommonFormsModule,
        ClrCombobox,
        ClrComboboxContainer,
        ClrOptions,
        ClrOption,
        ClrOptionSelected,
        ClrConditionalModule,
        ClrOptionItems]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrComboboxModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ClrIconModule,
                    ClrKeyFocusModule,
                    ClrCommonFormsModule,
                    ClrConditionalModule,
                    ClrPopoverModuleNext,
                    ClrSpinnerModule,
                ],
                declarations: [ClrCombobox, ClrComboboxContainer, ClrOptions, ClrOption, ClrOptionSelected, ClrOptionItems],
                exports: [
                    ClrCommonFormsModule,
                    ClrCombobox,
                    ClrComboboxContainer,
                    ClrOptions,
                    ClrOption,
                    ClrOptionSelected,
                    ClrConditionalModule,
                    ClrOptionItems,
                ]
            }]
    }], function () { return []; }, null); })();
