// @clr/angular/esm2015/forms/combobox/option-selected.directive.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, Input, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class ClrOptionSelected {
    constructor(template) {
        this.template = template;
    }
}
ClrOptionSelected.ɵfac = function ClrOptionSelected_Factory(t) { return new (t || ClrOptionSelected)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
ClrOptionSelected.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrOptionSelected, selectors: [["", "clrOptionSelected", ""]], inputs: { selected: ["clrOptionSelected", "selected"] } });
ClrOptionSelected.ctorParameters = () => [
    { type: TemplateRef }
];
ClrOptionSelected.propDecorators = {
    selected: [{ type: Input, args: ['clrOptionSelected',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrOptionSelected, [{
        type: Directive,
        args: [{ selector: '[clrOptionSelected]' }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, { selected: [{
            type: Input,
            args: ['clrOptionSelected']
        }] }); })();
