// angular-gridster2/esm5/lib/gridster.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { GridsterComponent } from './gridster.component';
import { GridsterItemComponent } from './gridsterItem.component';
import { GridsterPreviewComponent } from './gridsterPreview.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
var GridsterModule = /** @class */ (function () {
    function GridsterModule() {
    }
GridsterModule.ɵfac = function GridsterModule_Factory(t) { return new (t || GridsterModule)(); };
GridsterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: GridsterModule });
GridsterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ providers: [], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(GridsterModule, { declarations: function () { return [GridsterComponent,
        GridsterItemComponent,
        GridsterPreviewComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [GridsterComponent, GridsterItemComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(GridsterModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    GridsterComponent,
                    GridsterItemComponent,
                    GridsterPreviewComponent
                ],
                imports: [
                    CommonModule
                ],
                exports: [GridsterComponent, GridsterItemComponent],
                providers: [],
                bootstrap: []
            }]
    }], function () { return []; }, null); })();
ɵngcc0.ɵɵsetComponentScope(GridsterComponent, [ɵngcc1.NgForOf, ɵngcc1.NgStyle, GridsterPreviewComponent], []);
    return GridsterModule;
}());
export { GridsterModule };
