// @clr/angular/esm2015/utils/chocolate/oompa-loompa.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { ChangeDetectorRef, Directive } from '@angular/core';
import { WillyWonka } from './willy-wonka';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './willy-wonka';
export class OompaLoompa {
    // FIXME: Request Injector once we move to Angular 4.2+, it'll allow easier refactors
    constructor(cdr, willyWonka) {
        this.subscription = willyWonka.chocolate.subscribe(() => {
            if (this.latestFlavor !== this.flavor) {
                willyWonka.disableChocolateCheck = true;
                cdr.detectChanges();
                willyWonka.disableChocolateCheck = false;
            }
        });
    }
    ngAfterContentChecked() {
        this.latestFlavor = this.flavor;
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
}
OompaLoompa.ɵfac = function OompaLoompa_Factory(t) { return new (t || OompaLoompa)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.WillyWonka)); };
OompaLoompa.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: OompaLoompa });
OompaLoompa.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: WillyWonka }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OompaLoompa, [{
        type: Directive
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.WillyWonka }]; }, null); })();
