// @clr/angular/esm2015/utils/chocolate/willy-wonka.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive } from '@angular/core';
import { Subject } from 'rxjs';
/*
 * After a conversation with the Angular core team, it turns out we don't have much of a choice for our
 * declarative API, we need to fight against change detection and its one-way flow. This is
 * currently the least dirty solution to do what we want.
 *
 * Do not modify or even use this class unless you know exactly what you're doing.
 * It has the potential to trigger change detection loops or kill app performances.
 */
import * as ɵngcc0 from '@angular/core';
export class WillyWonka {
    constructor() {
        this._chocolate = new Subject();
        this.disableChocolateCheck = false;
    }
    get chocolate() {
        return this._chocolate.asObservable();
    }
    ngAfterViewChecked() {
        if (!this.disableChocolateCheck) {
            this._chocolate.next();
        }
    }
}
WillyWonka.ɵfac = function WillyWonka_Factory(t) { return new (t || WillyWonka)(); };
WillyWonka.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: WillyWonka });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WillyWonka, [{
        type: Directive
    }], function () { return []; }, null); })();
