// @clr/angular/esm2015/button/button-loading/loading-button.module.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrLoadingButton } from './loading-button';
import * as ɵngcc0 from '@angular/core';
export const CLR_LOADING_BUTTON_DIRECTIVES = [ClrLoadingButton];
export class ClrLoadingButtonModule {
}
ClrLoadingButtonModule.ɵfac = function ClrLoadingButtonModule_Factory(t) { return new (t || ClrLoadingButtonModule)(); };
ClrLoadingButtonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrLoadingButtonModule });
ClrLoadingButtonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrLoadingButtonModule, { declarations: function () { return [ClrLoadingButton]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrLoadingButton]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrLoadingButtonModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [CLR_LOADING_BUTTON_DIRECTIVES],
                exports: [CLR_LOADING_BUTTON_DIRECTIVES]
            }]
    }], null, null); })();
