// @clr/angular/esm2015/accordion/stepper/step-button.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, HostListener, HostBinding, Input } from '@angular/core';
import { StepperService } from './providers/stepper.service';
import { ClrStepperPanel } from './stepper-panel';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './stepper-panel';
import * as ɵngcc2 from './providers/stepper.service';
export var ClrStepButtonType;
(function (ClrStepButtonType) {
    ClrStepButtonType["Next"] = "next";
    ClrStepButtonType["Submit"] = "submit";
})(ClrStepButtonType || (ClrStepButtonType = {}));
export class ClrStepButton {
    constructor(clrStep, stepperService) {
        this.clrStep = clrStep;
        this.stepperService = stepperService;
        this.type = ClrStepButtonType.Next;
        this.submitButton = false;
    }
    ngOnInit() {
        this.submitButton = this.type === ClrStepButtonType.Submit;
    }
    navigateToNextPanel() {
        this.stepperService.navigateToNextPanel(this.clrStep.id, this.clrStep.formGroup.valid);
    }
}
ClrStepButton.ɵfac = function ClrStepButton_Factory(t) { return new (t || ClrStepButton)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ClrStepperPanel), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.StepperService)); };
ClrStepButton.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrStepButton, selectors: [["", "clrStepButton", ""]], hostVars: 7, hostBindings: function ClrStepButton_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ClrStepButton_click_HostBindingHandler() { return ctx.navigateToNextPanel(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("type", "button");
        ɵngcc0.ɵɵclassProp("clr-step-button", true)("btn", true)("btn-primary", ctx.submitButton);
    } }, inputs: { type: ["clrStepButton", "type"] } });
ClrStepButton.ctorParameters = () => [
    { type: ClrStepperPanel },
    { type: StepperService }
];
ClrStepButton.propDecorators = {
    type: [{ type: Input, args: ['clrStepButton',] }],
    submitButton: [{ type: HostBinding, args: ['class.btn-primary',] }],
    navigateToNextPanel: [{ type: HostListener, args: ['click',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrStepButton, [{
        type: Directive,
        args: [{
                selector: '[clrStepButton]',
                host: {
                    '[class.clr-step-button]': 'true',
                    '[class.btn]': 'true',
                    '[type]': "'button'"
                }
            }]
    }], function () { return [{ type: ɵngcc1.ClrStepperPanel }, { type: ɵngcc2.StepperService }]; }, { type: [{
            type: Input,
            args: ['clrStepButton']
        }], submitButton: [{
            type: HostBinding,
            args: ['class.btn-primary']
        }], navigateToNextPanel: [{
            type: HostListener,
            args: ['click']
        }] }); })();
