// @clr/angular/esm2015/accordion/stepper/stepper.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrAccordionModule } from '../accordion.module';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrStepButton } from './../stepper/step-button';
import { ClrStepper } from './../stepper/stepper';
import { ClrStepperPanel } from './stepper-panel';
import { StepperOompaLoompa } from './chocolate/stepper-oompa-loompa';
import { StepperWillyWonka } from './chocolate/stepper-willy-wonka';
import * as ɵngcc0 from '@angular/core';
const declarations = [ClrStepper, ClrStepButton, ClrStepperPanel, StepperOompaLoompa, StepperWillyWonka];
export class ClrStepperModule {
}
ClrStepperModule.ɵfac = function ClrStepperModule_Factory(t) { return new (t || ClrStepperModule)(); };
ClrStepperModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrStepperModule });
ClrStepperModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, ClrIconModule, ClrAccordionModule], ClrAccordionModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrStepperModule, { declarations: function () { return [ClrStepper, ClrStepButton, ClrStepperPanel, StepperOompaLoompa, StepperWillyWonka]; }, imports: function () { return [CommonModule, ClrIconModule, ClrAccordionModule]; }, exports: function () { return [ClrStepper, ClrStepButton, ClrStepperPanel, StepperOompaLoompa, StepperWillyWonka, ClrAccordionModule]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrStepperModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrAccordionModule],
                declarations: [...declarations],
                exports: [...declarations, ClrAccordionModule]
            }]
    }], null, null); })();
