// @clr/angular/esm2015/popover/signpost/signpost.module.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClarityIcons, infoCircleIcon, windowCloseIcon } from '@cds/core/icon';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrConditionalModule } from '../../utils/conditional/conditional.module';
import { ClrFocusOnViewInitModule } from '../../utils/focus/focus-on-view-init/focus-on-view-init.module';
import { ClrSignpost } from './signpost';
import { ClrSignpostContent } from './signpost-content';
import { ClrSignpostTrigger } from './signpost-trigger';
import * as ɵngcc0 from '@angular/core';
export const CLR_SIGNPOST_DIRECTIVES = [ClrSignpost, ClrSignpostContent, ClrSignpostTrigger];
export class ClrSignpostModule {
    constructor() {
        ClarityIcons.addIcons(windowCloseIcon, infoCircleIcon);
    }
}
ClrSignpostModule.ɵfac = function ClrSignpostModule_Factory(t) { return new (t || ClrSignpostModule)(); };
ClrSignpostModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrSignpostModule });
ClrSignpostModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, ClrIconModule, ClrFocusOnViewInitModule], ClrConditionalModule] });
ClrSignpostModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrSignpostModule, { declarations: function () { return [ClrSignpost, ClrSignpostContent, ClrSignpostTrigger]; }, imports: function () { return [CommonModule, ClrIconModule, ClrFocusOnViewInitModule]; }, exports: function () { return [ClrSignpost, ClrSignpostContent, ClrSignpostTrigger, ClrConditionalModule]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrSignpostModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrFocusOnViewInitModule],
                declarations: [CLR_SIGNPOST_DIRECTIVES],
                exports: [CLR_SIGNPOST_DIRECTIVES, ClrConditionalModule]
            }]
    }], function () { return []; }, null); })();
