// @clr/angular/esm2015/layout/tabs/tab.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, ContentChild, Inject } from '@angular/core';
import { IF_ACTIVE_ID, IF_ACTIVE_ID_PROVIDER, IfActiveService } from '../../utils/conditional/if-active.service';
import { AriaService } from '../../utils/aria/aria.service';
import { TabsService } from './providers/tabs.service';
import { ClrTabContent } from './tab-content';
import { ClrTabLink } from './tab-link.directive';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../utils/conditional/if-active.service';
import * as ɵngcc2 from './providers/tabs.service';

const _c0 = ["*"];
export class ClrTab {
    constructor(ifActiveService, id, tabsService) {
        this.ifActiveService = ifActiveService;
        this.id = id;
        this.tabsService = tabsService;
        tabsService.register(this);
    }
    ngOnDestroy() {
        this.tabsService.unregister(this);
    }
    get active() {
        return this.ifActiveService.current === this.id;
    }
}
ClrTab.ɵfac = function ClrTab_Factory(t) { return new (t || ClrTab)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.IfActiveService), ɵngcc0.ɵɵdirectiveInject(IF_ACTIVE_ID), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TabsService)); };
ClrTab.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrTab, selectors: [["clr-tab"]], contentQueries: function ClrTab_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ClrTabLink, 3);
        ɵngcc0.ɵɵcontentQuery(dirIndex, ClrTabContent, 3);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tabLink = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tabContent = _t.first);
    } }, features: [ɵngcc0.ɵɵProvidersFeature([IF_ACTIVE_ID_PROVIDER, AriaService])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrTab_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ClrTab.ctorParameters = () => [
    { type: IfActiveService },
    { type: Number, decorators: [{ type: Inject, args: [IF_ACTIVE_ID,] }] },
    { type: TabsService }
];
ClrTab.propDecorators = {
    tabLink: [{ type: ContentChild, args: [ClrTabLink, { static: true },] }],
    tabContent: [{ type: ContentChild, args: [ClrTabContent, { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrTab, [{
        type: Component,
        args: [{
                selector: 'clr-tab',
                template: ` <ng-content></ng-content> `,
                providers: [IF_ACTIVE_ID_PROVIDER, AriaService]
            }]
    }], function () { return [{ type: ɵngcc1.IfActiveService }, { type: Number, decorators: [{
                type: Inject,
                args: [IF_ACTIVE_ID]
            }] }, { type: ɵngcc2.TabsService }]; }, { tabLink: [{
            type: ContentChild,
            args: [ClrTabLink, { static: true }]
        }], tabContent: [{
            type: ContentChild,
            args: [ClrTabContent, { static: true }]
        }] }); })();
