// @clr/angular/esm2015/forms/checkbox/checkbox.module.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { checkCircleIcon, ClarityIcons, exclamationCircleIcon } from '@cds/core/icon';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrHostWrappingModule } from '../../utils/host-wrapping/host-wrapping.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrCheckbox } from './checkbox';
import { ClrCheckboxContainer } from './checkbox-container';
import { ClrCheckboxWrapper } from './checkbox-wrapper';
import * as ɵngcc0 from '@angular/core';
export class ClrCheckboxModule {
    constructor() {
        ClarityIcons.addIcons(exclamationCircleIcon, checkCircleIcon);
    }
}
ClrCheckboxModule.ɵfac = function ClrCheckboxModule_Factory(t) { return new (t || ClrCheckboxModule)(); };
ClrCheckboxModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrCheckboxModule });
ClrCheckboxModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, ClrIconModule, ClrCommonFormsModule, ClrHostWrappingModule], ClrCommonFormsModule] });
ClrCheckboxModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrCheckboxModule, { declarations: function () { return [ClrCheckbox, ClrCheckboxContainer, ClrCheckboxWrapper]; }, imports: function () { return [CommonModule, ClrIconModule, ClrCommonFormsModule, ClrHostWrappingModule]; }, exports: function () { return [ClrCommonFormsModule, ClrCheckbox, ClrCheckboxContainer, ClrCheckboxWrapper]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrCheckboxModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrCommonFormsModule, ClrHostWrappingModule],
                declarations: [ClrCheckbox, ClrCheckboxContainer, ClrCheckboxWrapper],
                exports: [ClrCommonFormsModule, ClrCheckbox, ClrCheckboxContainer, ClrCheckboxWrapper],
                entryComponents: [ClrCheckboxWrapper]
            }]
    }], function () { return []; }, null); })();
