// @cds/core/pagination/pagination.element.js
/*
 * Copyright (c) 2016-2021 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { baseStyles, LogService, notProductionEnvironment, querySlot } from '@cds/core/internal';
import { html, LitElement } from 'lit-element';
import { cache } from 'lit-html/directives/cache.js';
import { styles } from './pagination.element.css.js';
/**
 * Web component pagination.
 *
 * ```typescript
 * import '@cds/core/pagination/register.js';
 * ```
 *
 * ```html
 * <cds-pagination>
 *   <cds-pagination-button action="first" disabled></cds-pagination-button>
 *   <cds-pagination-button action="prev" disabled></cds-pagination-button>
 *   <span aria-label="current page">1 / 3</span>
 *   <cds-pagination-button action="next"></cds-pagination-button>
 *   <cds-pagination-button action="last"></cds-pagination-button>
 * </cds-pagination>
 * ```
 *
 * @element cds-pagination
 * @slot
 * @slot cds-pagination-button
 * @cssprop --color
 * @cssprop --font-size
 */
export class CdsPagination extends LitElement {
    render() {
        const userLayout = html `<slot></slot>`;
        const defaultLayout = html `<div cds-layout="horizontal gap:md align:center"><slot></slot></div>`;
        return html `${cache(this.hasAttribute('cds-layout') ? userLayout : defaultLayout)}`;
    }
    static get styles() {
        return [baseStyles, styles];
    }
    firstUpdated(props) {
        super.firstUpdated(props);
        this.setupNumberInput();
        this.validateAriaLabel();
    }
    setupNumberInput() {
        if (this.numberInput) {
            this.numberInput.controlWidth = 'shrink';
            this.numberInput.layout = 'compact';
        }
    }
    validateAriaLabel() {
        var _a;
        if (notProductionEnvironment() && !((_a = this.getAttribute('aria-label')) === null || _a === void 0 ? void 0 : _a.length)) {
            LogService.warn('An aria-label is missing', this);
        }
    }
}
__decorate([
    querySlot('cds-input[cds-pagination-number]')
], CdsPagination.prototype, "numberInput", void 0);
