// @clr/angular/esm2015/accordion/chocolate/accordion-oompa-loompa.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { ChangeDetectorRef, Directive, Optional } from '@angular/core';
import { OompaLoompa } from '../../utils/chocolate/oompa-loompa';
import { AccordionWillyWonka } from './accordion-willy-wonka';
import { IfExpandService } from '../../utils/conditional/if-expanded.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './accordion-willy-wonka';
import * as ɵngcc2 from '../../utils/conditional/if-expanded.service';
export class AccordionOompaLoompa extends OompaLoompa {
    constructor(cdr, willyWonka, ifExpandService) {
        if (!willyWonka) {
            throw new Error('clr-accordion-panel should only be used inside of clr-accordion');
        }
        super(cdr, willyWonka);
        this.expand = ifExpandService;
    }
    get flavor() {
        return this.expand.expanded;
    }
}
AccordionOompaLoompa.ɵfac = function AccordionOompaLoompa_Factory(t) { return new (t || AccordionOompaLoompa)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AccordionWillyWonka, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.IfExpandService)); };
AccordionOompaLoompa.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AccordionOompaLoompa, selectors: [["clr-accordion-panel"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
AccordionOompaLoompa.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: AccordionWillyWonka, decorators: [{ type: Optional }] },
    { type: IfExpandService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AccordionOompaLoompa, [{
        type: Directive,
        args: [{ selector: 'clr-accordion-panel' }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.AccordionWillyWonka, decorators: [{
                type: Optional
            }] }, { type: ɵngcc2.IfExpandService }]; }, null); })();
