// @clr/angular/esm2015/timeline/timeline.module.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { circleIcon, ClarityIcons, dotCircleIcon, errorStandardIcon, successStandardIcon } from '@cds/core/icon';
import { ClrIconModule } from '../icon/icon.module';
import { ClrSpinnerModule } from '../progress/spinner/spinner.module';
import { ClrTimeline } from './timeline';
import { ClrTimelineStep } from './timeline-step';
import { ClrTimelineStepDescription } from './timeline-step-description';
import { ClrTimelineStepHeader } from './timeline-step-header';
import { ClrTimelineStepTitle } from './timeline-step-title';
import * as ɵngcc0 from '@angular/core';
const CLR_TIMELINE_DIRECTIVES = [
    ClrTimeline,
    ClrTimelineStep,
    ClrTimelineStepDescription,
    ClrTimelineStepHeader,
    ClrTimelineStepTitle,
];
export class ClrTimelineModule {
    constructor() {
        ClarityIcons.addIcons(circleIcon, dotCircleIcon, errorStandardIcon, successStandardIcon);
    }
}
ClrTimelineModule.ɵfac = function ClrTimelineModule_Factory(t) { return new (t || ClrTimelineModule)(); };
ClrTimelineModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrTimelineModule });
ClrTimelineModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, ClrIconModule, ClrSpinnerModule], ClrIconModule, ClrSpinnerModule] });
ClrTimelineModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrTimelineModule, { declarations: function () { return [ClrTimeline,
        ClrTimelineStep,
        ClrTimelineStepDescription,
        ClrTimelineStepHeader,
        ClrTimelineStepTitle]; }, imports: function () { return [CommonModule, ClrIconModule, ClrSpinnerModule]; }, exports: function () { return [ClrTimeline,
        ClrTimelineStep,
        ClrTimelineStepDescription,
        ClrTimelineStepHeader,
        ClrTimelineStepTitle, ClrIconModule, ClrSpinnerModule]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrTimelineModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrSpinnerModule],
                exports: [...CLR_TIMELINE_DIRECTIVES, ClrIconModule, ClrSpinnerModule],
                declarations: [CLR_TIMELINE_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
