// @clr/angular/esm2015/utils/popover/popover.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 *
 */
import { NgModule } from '@angular/core';
import { ClrPopoverContent } from './popover-content';
import { ClrPopoverOpenCloseButton } from './popover-open-close-button';
import { ClrPopoverCloseButton } from './popover-close-button';
import { ClrPopoverAnchor } from './popover-anchor';
import * as ɵngcc0 from '@angular/core';
export class ClrPopoverModuleNext {
}
ClrPopoverModuleNext.ɵfac = function ClrPopoverModuleNext_Factory(t) { return new (t || ClrPopoverModuleNext)(); };
ClrPopoverModuleNext.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrPopoverModuleNext });
ClrPopoverModuleNext.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrPopoverModuleNext, { declarations: function () { return [ClrPopoverAnchor, ClrPopoverCloseButton, ClrPopoverOpenCloseButton, ClrPopoverContent]; }, exports: function () { return [ClrPopoverAnchor, ClrPopoverCloseButton, ClrPopoverOpenCloseButton, ClrPopoverContent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrPopoverModuleNext, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [ClrPopoverAnchor, ClrPopoverCloseButton, ClrPopoverOpenCloseButton, ClrPopoverContent],
                exports: [ClrPopoverAnchor, ClrPopoverCloseButton, ClrPopoverOpenCloseButton, ClrPopoverContent]
            }]
    }], null, null); })();
