// @clr/angular/esm2015/utils/popover/popover-anchor.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 *
 */
import { Directive, ElementRef } from '@angular/core';
import { ClrPopoverEventsService } from './providers/popover-events.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './providers/popover-events.service';
export class ClrPopoverAnchor {
    constructor(smartEventService, element) {
        smartEventService.anchorButtonRef = element;
    }
}
ClrPopoverAnchor.ɵfac = function ClrPopoverAnchor_Factory(t) { return new (t || ClrPopoverAnchor)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ClrPopoverEventsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ClrPopoverAnchor.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrPopoverAnchor, selectors: [["", "clrPopoverAnchor", ""]], hostVars: 2, hostBindings: function ClrPopoverAnchor_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-anchor", true);
    } } });
ClrPopoverAnchor.ctorParameters = () => [
    { type: ClrPopoverEventsService },
    { type: ElementRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrPopoverAnchor, [{
        type: Directive,
        args: [{
                selector: '[clrPopoverAnchor]',
                host: {
                    '[class.clr-anchor]': 'true'
                }
            }]
    }], function () { return [{ type: ɵngcc1.ClrPopoverEventsService }, { type: ɵngcc0.ElementRef }]; }, null); })();
