// @swimlane/ngx-charts/esm2015/lib/box-chart/box-chart.module.js
import { NgModule } from '@angular/core';
import { ChartCommonModule } from '../common/chart-common.module';
import { BoxChartComponent } from './box-chart.component';
import { BoxSeriesComponent } from './box-series.component';
import { BoxComponent } from './box.component';
import * as ɵngcc0 from '@angular/core';
export class BoxChartModule {
}
BoxChartModule.ɵfac = function BoxChartModule_Factory(t) { return new (t || BoxChartModule)(); };
BoxChartModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: BoxChartModule });
BoxChartModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[ChartCommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BoxChartModule, { declarations: function () { return [BoxChartComponent, BoxSeriesComponent, BoxComponent]; }, imports: function () { return [ChartCommonModule]; }, exports: function () { return [BoxChartComponent, BoxSeriesComponent, BoxComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BoxChartModule, [{
        type: NgModule,
        args: [{
                imports: [ChartCommonModule],
                declarations: [BoxChartComponent, BoxSeriesComponent, BoxComponent],
                exports: [BoxChartComponent, BoxSeriesComponent, BoxComponent]
            }]
    }], null, null); })();
