// @clr/angular/esm2015/progress/progress-bars/progress-bar.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrProgressBar } from './progress-bar';
import * as ɵngcc0 from '@angular/core';
export const CLR_PROGRESS_BAR_DIRECTIVES = [ClrProgressBar];
export class ClrProgressBarModule {
}
ClrProgressBarModule.ɵfac = function ClrProgressBarModule_Factory(t) { return new (t || ClrProgressBarModule)(); };
ClrProgressBarModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrProgressBarModule });
ClrProgressBarModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrProgressBarModule, { declarations: function () { return [ClrProgressBar]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrProgressBar]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClrProgressBarModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [CLR_PROGRESS_BAR_DIRECTIVES],
                exports: [CLR_PROGRESS_BAR_DIRECTIVES]
            }]
    }], null, null); })();
